scripts/azure-price-fetch/azureClientWrapper.go

```diff
@@ -0,0 +1,177 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v4"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/Azure/go-autorest/autorest/to"
+	"gomodules.xyz/azure-retail-prices-sdk-for-go/sdk"
+)
+
+type AzureClientWrapper struct {
+	subscriptionId string
+	priceClient    *sdk.RetailPricesClient
+	rgClient       *armresources.ResourceGroupsClient
+	vmssClient     *armcompute.VirtualMachineScaleSetsClient
+	vmssVmClient   *armcompute.VirtualMachineScaleSetVMsClient
+}
+
+func NewAzureClientWrapper(subId string, cred *azidentity.DefaultAzureCredential) (*AzureClientWrapper, error) {
+	retailPricesClient, err := sdk.NewRetailPricesClient(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create retail prices client: %w", err)
+	}
+
+	rgClient, err := armresources.NewResourceGroupsClient(subId, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build resource group client: %w", err)
+	}
+
+	computeClientFactory, err := armcompute.NewClientFactory(subId, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build compute client: %w", err)
+	}
+
+	return &AzureClientWrapper{
+		subscriptionId: subId,
+
+		priceClient:  retailPricesClient,
+		rgClient:     rgClient,
+		vmssClient:   computeClientFactory.NewVirtualMachineScaleSetsClient(),
+		vmssVmClient: computeClientFactory.NewVirtualMachineScaleSetVMsClient(),
+	}, nil
+}
+
+func (a *AzureClientWrapper) buildQueryFilter(locationList []string) string {
+	locationListFilter := []string{}
+	for _, t := range locationList {
+		locationListFilter = append(locationListFilter, fmt.Sprintf("armRegionName eq '%s'", t))
+	}
+
+	locationListStr := strings.Join(locationListFilter, " or ")
+	return fmt.Sprintf(`serviceName eq 'Virtual Machines' and priceType eq 'Consumption' and (%s)`, locationListStr)
+}
+
+func (a *AzureClientWrapper) getPrices(ctx context.Context, locationList []string) (*PriceMap, error) {
+	pm := PriceMap{
+		RegionMap: make(map[string]map[string]sdk.ResourceSKU),
+	}
+
+	pager := a.priceClient.NewListPager(&sdk.RetailPricesClientListOptions{
+		APIVersion:  to.StringPtr("2023-01-01-preview"),
```

Comment:
What's the significance of 2023-01-01-preview? I'd suggest moving it to a const and possibly link out to what it implies.

Reply:
updated